kroxylicious-filters/kroxylicious-oauthbearer-validation/src/main/java/io/kroxylicious/proxy/filter/oauthbearer/OauthBearerValidationFilter.java

```diff
@@ -0,0 +1,212 @@
+                        }
+                        else {
+                            LOGGER.debug("SASL error : {}", e.getMessage(), e);
+                            return context.requestFilterResultBuilder()
+                                    .shortCircuitResponse(
+                                            new SaslAuthenticateResponseData()
+                                                    .setErrorCode(UNKNOWN_SERVER_ERROR.code())
+                                                    .setAuthBytes(request.authBytes()))
+                                    .withCloseConnection()
+                                    .completed();
+                        }
+                    });
+        }
+    }
+
+    @Override
+    public CompletionStage<ResponseFilterResult> onSaslAuthenticateResponse(short apiVersion, ResponseHeaderData header,
+                                                                            SaslAuthenticateResponseData response, FilterContext context) {
+        if (response.errorCode() == NONE.code()) {
+            this.validateAuthentication = false;
+        }
+        return context.forwardResponse(header, response);
+    }
+
+    private CompletionStage<byte[]> authenticate(byte[] authBytes) {
+        String digest;
+        try {
+            digest = digestBytes(authBytes);
+        }
+        catch (NoSuchAlgorithmException e) {
+            return CompletableFuture.failedStage(e);
+        }
+        Duration delay = strategy.getDelay(rateLimiter.get(digest).get());
+        return schedule(() -> {
+            try {
+                return CompletableFuture.completedStage(doAuthenticate(authBytes));
+            }
+            catch (Exception e) {
+                return CompletableFuture.failedStage(e);
+            }
+        }, delay)
+                .whenComplete((bytes, e) -> {
+                    if (e != null) {
+                        rateLimiter.get(digest).incrementAndGet();
+                    }
+                    else {
+                        rateLimiter.invalidate(digest);
+                    }
+                });
+    }
+
+    private byte[] doAuthenticate(byte[] authBytes) throws SaslException {
+        try {
+            byte[] bytes = this.saslServer.evaluateResponse(authBytes);
+            if (!this.saslServer.isComplete()) {
+                throw new SaslAuthenticationException("SASL failed : " + new String(bytes, StandardCharsets.UTF_8));
+            }
+            return bytes;
+        }
+        finally {
+            this.saslServer.dispose();
+        }
+    }
+
+    @SuppressWarnings("java:S1602") // not able to test the scheduled lambda otherwise
+    private <A> CompletionStage<A> schedule(Supplier<CompletionStage<A>> operation, Duration duration) {
+        if (duration.equals(Duration.ZERO)) {
+            return operation.get();
+        }
+        CompletableFuture<A> future = new CompletableFuture<>();
+        executorService.schedule(() -> {
```

Comment:
By discarding the returned `ScheduledFuture` you're basically allowing the client to send very many `SaslAuthenticateRequests`, each with its own scheduled completion. Other clients are sharing the same event loop (which is what the `executorService` here actually is), so with enough scheduled events you might get an obserable noisy neighbour effect.

We should probably hold the last scheduled operation's `ScheduledFuture` in a field. That would allow us to implement a one-scheduler-slot-per-client policy by cancelling the outstanding operation. We'd need to decide what the right thing to do with the two in-flight requests is. I guess respond to the first with some error code, and allow the second to be scheduled.

Reply:
If we cancel the first request, a response will be sent to the client no ? How this behavior will prevent the client to start new requests ? We can effectively limit the number of scheduled operations to not , but I don't see how this helps to prevent too many requests ?
Maybe we can refuse new requests from client when another one is already scheduled to fast short circuit ?